tests/test_extension.py

```diff
@@ -0,0 +1,683 @@
+        """Test if the existing extension implemented the interface with
+        less arguments
+        """
+        class ExtensionWithLessArgument(FuncExtensionBase):
+            def __init__(self, file_path: str):
+                super().__init__(file_path)
+                self.executed = False
+
+            # Drop arguments
+            def pre_invocation(self):
+                self.executed = True
+
+        # Check if the before invocation hook matches metadata
+        ext_instance = ExtensionWithLessArgument(self.mock_file_path)
+        hooks = ExtensionMeta.get_function_hooks('HttpTrigger')
+
+        # Check if implementation works
+        hook_meta = hooks.pre_invocation[0]
+        self.assertEqual(hook_meta.ext_name, 'ExtensionWithLessArgument')
+
+        # Check if the hook implementation executes
+        hook_meta.ext_impl()
+        self.assertTrue(ext_instance.executed)
+
+
+class TestAppExtensionBase(unittest.TestCase):
+
+    def setUp(self):
+        self.patch_os_environ = patch.dict('os.environ', os.environ.copy())
+        self.patch_os_environ.start()
+
+    def tearDown(self) -> None:
+        self.patch_os_environ.stop()
+        ExtensionMeta._info.clear()
+        ExtensionMeta._func_exts.clear()
+        ExtensionMeta._app_exts = None
+
+    def test_empty_extension_should_pass(self):
+        """An application extension can be registered directly since it never
+        gets instantiate
+        """
+        class NewAppExtension(AppExtensionBase):
+            pass
+
+    def test_init_method_should_be_called(self):
+        """An application extension's init() classmethod should be called
+        when the class is created"""
+        class NewAppExtension(AppExtensionBase):
+            _initialized = False
+
+            @classmethod
+            def init(cls):
+                cls._initialized = True
+
+        self.assertTrue(NewAppExtension._initialized)
+
+    def test_extension_registration(self):
+        """The life-cycles implementations in extension should be automatically
+        registered in class creation
+        """
+        class NewAppExtension(AppExtensionBase):
+            @classmethod
+            def post_function_load_app_level(cls,
+                                             function_name,
+                                             function_directory,
+                                             *args,
+                                             **kwargs) -> None:
+                print('ok_post_function_load_app_level')
+
+            @classmethod
+            def pre_invocation_app_level(self, logger, context,
+                                         *args, **kwargs) -> None:
+                logger.info('ok_pre_invocation_app_level')
+
+            @classmethod
+            def post_invocation_app_level(self, logger, context,
+                                          *args, **kwargs) -> None:
+                logger.info('ok_post_invocation_app_level')
```

Comment:
Would returning a string and checking the returned string be better than logging?